allennlp/models/multitask.py

```diff
@@ -84,58 +84,59 @@ def __init__(
         super().__init__(vocab, **kwargs)
         self._backbone = backbone
         self._heads = torch.nn.ModuleDict(heads)
+        self._heads_called: Set[str] = set()
         self._arg_name_mapping = arg_name_mapping or defaultdict(dict)
 
         self._allowed_arguments = allowed_arguments or {
             "backbone": get_forward_arguments(backbone),
             **{key: get_forward_arguments(heads[key]) for key in heads},
         }
         self._loss_weights = loss_weights or defaultdict(lambda: 1.0)
-        self._active_heads: Optional[List[str]] = None
         initializer(self)
 
-    def set_active_heads(self, active_heads: List[str]) -> None:
-        """
-        By default, the `MultiTaskModel` will try to infer which heads to run from the arguments
-        passed to `forward`.  During training, we will only run a head if we have all of its
-        arguments, including optional arguments, which typically means the argument is the
-        prediction target; if we don't have it, we typically can't compute a loss, so running during
-        training is pointless.  During evaluation, we will run all heads.
-
-        If you want to limit which heads are run during evaluation, or if the inference for which
-        task to run during training is incorrect (e.g., if your head has multiple optional
-        arguments, and only some are actually required to compute a loss), then you can use this
-        method to override our inference and force the use of whatever set of heads you want.
-
-        To get back to the default mode of operation, call this method with `None` as an argument.
-        """
-        self._active_heads = active_heads
-
     def forward(self, **kwargs) -> Dict[str, torch.Tensor]:  # type: ignore
+        if "task" not in kwargs:
```

Comment:
The change here is that the old code automatically detected which heads to run based on the input parameters. If the signature fits, it would run. In the case of VQA/GQA though, we are running exactly the same head twice, once for each dataset, so this approach couldn't work. I changed it so now the instance keeps the task name in a `MetadataField`, and uses that to determine which task to run.